parser/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g

```diff
@@ -1840,6 +1841,14 @@ tableImplBuckets
     -> ^(TOK_ALTERTABLE_BUCKETS $num)
     ;
 
+tableWriteOrdered
+@init { pushMsg("table sorted specification", state); }
+@after { popMsg(state); }
+    :
+      KW_WRITE KW_ORDERED KW_BY sortCols=columnNameOrderList
```

Comment:
Does this syntax equal to Spark-Iceberg ` alter-table-write-ordered-by`?
https://iceberg.apache.org/docs/1.7.0/spark-ddl/#alter-table-write-ordered-by

Do you consider also to add the ` alter-table-write-ordered-by` syntax? So that user can alter the existing iceberg table to support write order.

Reply:
BTW, Trino added the `write-sort ` through table property:
https://trino.io/docs/current/connector/iceberg.html#sorted-tables
https://www.starburst.io/blog/improving-performance-with-iceberg-sorted-tables/


```
CREATE TABLE
catalog_sales_sorted (
   cs_sold_date_sk bigint,
   more columns...
)
WITH
(
   format = 'PARQUET',
   sorted_by = ARRAY['cs_sold_date_sk'],
   type = 'ICEBERG'
)
```

This can avoid introducing new create syntax. Anyway, we should find a simple syntax which user can use.

I also want to hear other folks opinions about the syntax. 
cc @ayushtkn @deniskuzZ @okumin

Reply:
Is the semantics of the sort order of Iceberg different from that of Hive's SORTED BY of bucketed sorted tables? If it is the same, we may use it. Maybe, it is different. I also feel the global ordering is hard to achieve with the current Hive.

Reply:
https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL
````
The CLUSTERED BY and SORTED BY creation commands do not affect how data is inserted into a table – only how it is read. This means that users must be careful to insert data correctly by specifying the number of reducers to be equal to the number of buckets, and using CLUSTER BY and SORT BY commands in their query.
````
@okumin, this PR intends to add support for Iceberg sorted tables, same as in a trino link you've added. I do not really like that we have to create new syntax for that. 
`SORTED BY` kinda makes sense, but it would be overloaded for Iceberg since it would be used to specify data insert order.

Spark only supports [alter-table-write-ordered-by](https://iceberg.apache.org/docs/1.7.0/spark-ddl/#alter-table-write-ordered-by) to define the write order. 

Maybe we could have a combination of ALTER + Trino table properties styles?

Reply:
I understand the SORT order of the Iceberg is a hint in writing.
- [Table Spec](https://iceberg.apache.org/spec/#sorting) says, "Writers should use this default sort order to sort the data on write, but are not required to if the default order is prohibitively expensive, as it would be for streaming writes."
- [Spark DDL](https://learn.microsoft.com/en-us/cli/azure/role/definition?view=azure-cli-latest#az-role-definition-list) says, "Table write order does not guarantee data order for queries. It only affects how data is written to the table."
- [Trino has a toggle, sorted_writing_enabled, to obey the hint](https://trino.io/docs/current/connector/iceberg.html#iceberg-sorted-files)

So, I agree that reusing SORTED BY is confusing. I'm also checking what syntaxes are used in other engines...

Reply:
## How other Iceberg-friendly query engines support SORT orders

### ALTER TABLE
- [Spark supports WRITE ORDERED BY and WRITE LOCALLY ORDERED BY](https://iceberg.apache.org/docs/1.7.0/spark-ddl/#alter-table-write-ordered-by)
- [Trino supports it through table properties](https://trino.io/docs/current/connector/iceberg.html#alter-table-set-properties)
- [Dremio supports LOCALSORT BY](https://docs.dremio.com/cloud/reference/sql/commands/alter-table)

### CREATE TABLE
- [Spark doesn't support it?](https://github.com/apache/iceberg/blob/apache-iceberg-1.7.1/spark/v3.5/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4)
- [Trino supports it through table properties](https://trino.io/docs/current/connector/iceberg.html#sorted-tables)
- [Dremio supports LOCALSORT BY](https://docs.dremio.com/cloud/reference/sql/commands/create-table/)

## Related discussions

I found a related discussion although it would not include the conclusion we want.
https://github.com/apache/iceberg/issues/3547

## My current opinion

If we introduce `ALTER TABLE ... WRITE ORDERED BY`, we can reuse it in CREATE TABLE in my opinion. It is consistent and no new keywords are introduced for CREATE TABLE. If we don't, I prefer to use a property for both purposes.

Reply:
@zhangbutao , @okumin , @deniskuzZ 
First of all, thanks for your comments! This syntax comes from the official Iceberg recommendation, which is for alter table, and to be consistend I modified the ALTER to CREATE in the syntax, so at CREATE table it is the same as ALTER TABLE:
[https://iceberg.apache.org/docs/1.7.0/spark-ddl/#alter-table-write-ordered-by]

My plan is to work as a next step on ALTER TABLE after this PR is done. I already created the ticket for that at the time this ticket was born: 
[https://issues.apache.org/jira/browse/HIVE-28587]

About table property idea and comparisms: 
The iceberg syntax defines multiple keywords in the above page, not just the colum names:

"ALTER TABLE prod.db.sample WRITE ORDERED BY category ASC NULLS LAST, id DESC NULLS FIRST"

The mentioned examples do not cover all of that. E.g. I could not find in the linked Trino documentation how the ordering is defined (ASC/DESC) or the null ordering (NULLS FIRST/LAST). With that the table property solution would look like ugly and hard to remember and write correctly for the user. 

Since it is passedas a table property deep in the code here, it works with table property as well with this commit :
e.g.:
create table ice_orc_sorted (id int, text string) stored by iceberg stored as orc TBLPROPERTIES('iceberg.write-order'='{"order-id":1,"fields":[{"transform":"identity","source-id":1,"direction":"asc","null-order":"nulls-last"}]}');

The syntax comes from Iceberg built in format for sorted order, but I do not find it user friendly and I do not recommend to go that way.


About SORTED BY/ SORT BY/LOCALSORT BY: as you already mentioned it would be confusing, and the "ALTER/CREATE TABLE prod.db.sample WRITE ORDERED BY category ASC NULLS LAST, id DESC NULLS FIRST" is recommended by Iceberg, and easy to understand and remember.

So every DB engine has it's own syntax for this purpose and **neither is following Iceberg recommendation except this one.**

As far as I found Spark does not even support yet sort order at alter table:
https://spark.apache.org/docs/3.5.3/sql-ref-syntax-ddl-alter-table.html#content
https://spark.apache.org/docs/3.5.3/sql-ref-syntax-ddl-create-table.html


At the linked iceberg discussion even Ryan Blue said this:
"No, I don't think so. These aren't table properties... "
And someone else said:
" I don't think using TBLPROPERTIES is a good idea..."

I hope I answered all the questions!

Reply:
I like the `LOCALSORT BY` it gives more context than `WRITE ORDERED BY`

Reply:
@zhangbutao @okumin @deniskuzZ I would like to finish this PR. Could you please give feedback if it is acceptable as I described the design of this ticket?

Reply:
> The mentioned examples do not cover all of that. E.g. I could not find in the linked Trino documentation how the ordering is defined (ASC/DESC) or the null ordering (NULLS FIRST/LAST)

I think this is a critical point, and now I am biased toward to `WRITE ORDERED BY` + `WRITE LOCALLY ORDERED BY`. I am trying to talk to a Trino maintainers

Reply:
JFYI, I heard Trino would support some detailed properties and verified the behavior.

```
trino> CREATE TABLE iceberg.default.test1(a int) WITH (sorted_by = ARRAY['a']);
CREATE TABLE
trino> CREATE TABLE iceberg.default.test2(a int) WITH (sorted_by = ARRAY['a DESC NULLS FIRST']);
CREATE TABLE
```

```
$ hdfs dfs -cat /user/hive/warehouse/test1-d38df51e56ee445aa4b100fc6306cd78/metadata/00000-77ea73cf-95ec-44fc-bf47-2287e19a4195.metadata.json
...
  "default-sort-order-id" : 1,
  "sort-orders" : [ {
    "order-id" : 1,
    "fields" : [ {
      "transform" : "identity",
      "source-id" : 1,
      "direction" : "asc",
      "null-order" : "nulls-first"
    } ]
  } ],
```

```
$ hdfs dfs -cat /user/hive/warehouse/test2-1d2d11372fad4f3e945d298f63afe322/metadata/00000-62c0c21b-2032-4883-a092-012bd1572e83.metadata.json
...
  "default-sort-order-id" : 1,
  "sort-orders" : [ {
    "order-id" : 1,
    "fields" : [ {
      "transform" : "identity",
      "source-id" : 1,
      "direction" : "desc",
      "null-order" : "nulls-first"
    } ]
  } ],
```

Reply:
First of all, I think the consistency between ALTER TABLE and CREATE TABLE is the most important.
On top of that, I think Spark is the most common engine for Iceberg. I'm reading O'Reilly's "Apache Iceberg: The Definitive Guide", and the first example of any feature is almost always Spark SQL. My preference here is to use the syntax introduced for Spark.

Reply:
@zratkai you have created the ticket [HIVE-28587](https://issues.apache.org/jira/browse/HIVE-28587), which means that you will implement the `alter table `syntax like spark syntax. https://iceberg.apache.org/docs/1.7.0/spark-ddl/#alter-table-write-ordered-by  Right?

> This syntax comes from the official Iceberg recommendation, which is for alter table, and to be consistend I modified the ALTER to CREATE in the syntax, so at CREATE table it is the same as ALTER TABLE:

I already know why you give this create syntax. But since there is no uniform `Create `syntax for the `write order`, i think we no need to make `Create `syntax to be consistend with the `Alter `syntax. If the table property is more simple, i prefer to `create + table property` like trino.

If you think `create + table property` is also not friendly to users. I am not against the currrent implementation. You just need to update the wiki doc to tell users the create dialect for iceberg write order.

Reply:
If there is no common agreement, We can reach out the dev mailing list & can proceed with what majority of the folks like, This Syntax discussion is kind of pure personal preference stuff I believe, what each of us find more convenient & there is no way we can conclude like this which is better.

Was reading this doc:
https://www.tabular.io/apache-iceberg-cookbook/data-engineering-table-write-order/
It had something like
```
ALTER TABLE logs WRITE LOCALLY ORDERED BY level, event_ts
```
& 

here:
https://iceberg.apache.org/docs/latest/spark-ddl/#alter-table-write-ordered-by

```
ALTER TABLE prod.db.sample WRITE LOCALLY ORDERED BY category, id
```

If LOCALLY isn't specified it is Global Sort order as the doc says 'To order within each task, not across tasks, use LOCALLY ORDERED BY:' and I believe this PR is just supporting LOCAL sort order not the Global one.

So, this kind of looks fine to me, We can just suffix in the Create command as well something like "WRITE LOCALLY ORDERED BY level, event_ts" which will be in sync with ALTER command & all such above docs.

Again just my preference, if all of us aren't convinced on one choice, we can fallback to dev@ and go what the majority feels more soothing to eyes.

Reply:
@ayushtkn thanks Ayush for the comment!
I agree with you and @deniskuzZ,  LOCALLY ORDERED BY is the right choice here.

@okumin @zhangbutao if you can accept that I will modify it accordingly.

Reply:
It's fine to me. 😃 Just do it. We can optimize this later if someone has new good idea in the future.

Reply:
I give +1 to WRITE LOCALLY ORDERED BY.

I double-checked `WRITE ORDERED BY` vs `WRITE LOCALLY ORDERED BY` with Spark 3.5.1 and Iceberg 1.6.1.

```
spark-sql (default)> CREATE TABLE hadoop_prod.default.test2 (a int) USING iceberg;
Time taken: 0.089 seconds
spark-sql (default)> ALTER TABLE hadoop_prod.default.test2 WRITE ORDERED BY a;
Time taken: 0.182 seconds
spark-sql (default)> CREATE TABLE hadoop_prod.default.test3 (a int) USING iceberg;
Time taken: 0.086 seconds
spark-sql (default)> ALTER TABLE hadoop_prod.default.test3 WRITE LOCALLY ORDERED BY a;
```

This is the diff.

```
zookage@client-node-0:~$ hdfs dfs -cat /user/hive/warehouse/catalog/default/test2/metadata/v2.metadata.json > /tmp/test2.json
zookage@client-node-0:~$ hdfs dfs -cat /user/hive/warehouse/catalog/default/test3/metadata/v2.metadata.json > /tmp/test3.json
zookage@client-node-0:~$ diff /tmp/test2.json /tmp/test3.json 
3,4c3,4
<   "table-uuid" : "821c3cc2-1320-45dc-bb2a-c805778caa91",
<   "location" : "hdfs://hdfs-namenode-0.hdfs-namenode:8020/user/hive/warehouse/catalog/default/test2",
---
>   "table-uuid" : "094951f1-6229-43cd-bffe-cb2f086b8dda",
>   "location" : "hdfs://hdfs-namenode-0.hdfs-namenode:8020/user/hive/warehouse/catalog/default/test3",
6c6
<   "last-updated-ms" : 1733994463428,
---
>   "last-updated-ms" : 1733994491871,
40c40
<     "write.distribution-mode" : "range",
---
>     "write.distribution-mode" : "none",
50,51c50,51
<     "timestamp-ms" : 1733994460005,
<     "metadata-file" : "hdfs://hdfs-namenode-0.hdfs-namenode:8020/user/hive/warehouse/catalog/default/test2/metadata/v1.metadata.json"
---
>     "timestamp-ms" : 1733994472725,
>     "metadata-file" : "hdfs://hdfs-namenode-0.hdfs-namenode:8020/user/hive/warehouse/catalog/default/test3/metadata/v1.metadata.json"
```

Looks like, the meaningful difference is only `write.distribution-mode=range` or `write.distribution-mode=none`. I guess adding LOCALLY makes more sense unless we give `range`.

I am not a specialist of Apache Spark. Please feel free to correct me if I am wrong.